modules/rapid/src/rapid.cpp

```diff
@@ -355,5 +355,46 @@ float rapid(InputArray img, int num, int len, InputArray vtx, InputArray tris, I
     return float(pts2d.rows) / num;
 }
 
+Tracker::~Tracker() {}
+
+struct RapidImpl : public Rapid
+{
+    Mat pts3d;
+    Mat tris;
+    RapidImpl(InputArray _pts3d, InputArray _tris)
+    {
+        CV_Assert(_tris.getMat().checkVector(3, CV_32S) > 0);
+        CV_Assert(_pts3d.getMat().checkVector(3, CV_32F) > 0);
+        pts3d = _pts3d.getMat();
+        tris = _tris.getMat();
+    }
+    float compute(InputArray img, int num, int len, InputArray K, InputOutputArray rvec,
+                  InputOutputArray tvec, const TermCriteria& termcrit) CV_OVERRIDE
+    {
+        float ret = 0;
+        int niter = std::max(1, termcrit.maxCount);
+
+        Mat cols;
+        for(int i = 0; i < niter; i++)
+        {
+            ret = rapid(img, num, len, pts3d, tris, K, rvec, tvec, cols);
+
+            if(termcrit.type & TermCriteria::EPS)
+            {
+                cols -= len + 1;
+                double rmsd = std::sqrt(norm(cols, NORM_L2SQR) / cols.rows);
```

Comment:
for clarification: this effectively measures, how far the object moved in 2D. so we always need at least 2 iterations to converge?

Reply:
since we assume an initial position, one iteration is enough to measure the 2D distance from that.

Reply:
I don't get it :)

In the following assume that moving one world unit maps to moving one pixel. The object moved 10px to the right in image space, no rotation.

Iteration one:
- init pose: t=(0, 0, 0)
- rapid does it stuff and finds a the optimal solution: t=(10, 0, 0)
- RMSD=10 => do another iteration
Iteration two:
- init pose: t=(10, 0, 0)
- rapid does it stuff and finds a the optimal solution: t=(10, 0, 0)
- RMSD=0 => finished

Reply:
yes, you always do one iteration more then you need. But with init pose: t=(10, 0, 0), you stop after 1 iteration